src/lib/transition_frontier/persistent_frontier/worker.ml

```diff
@@ -4,21 +4,78 @@ open Otp_lib
 open Coda_base
 open Frontier_base
 
+type input = Diff.Lite.E.t list
+
+type create_args = {db: Database.t; logger: Logger.t}
+
 module Worker = struct
   (* when this is transitioned to an RPC worker, the db argument
    * should just be a directory, but while this is still in process,
    * the full instance is needed to share with other threads
    *)
-  type t = {db: Database.t; logger: Logger.t}
+  type nonrec input = input
 
-  type create_args = t
+  type t =
+    { db: Database.t
+    ; logger: Logger.t
+          (* Invariant:
 
-  type input = Diff.Lite.E.t list * Frontier_hash.t
+   If there are root transitions in this queue which can be performed, we perform
+   them immediately. This invariant is potentially violated whenever we touch the
+   database (i.e., whenever we process any other kind of diff) so we eagerly perform
+   work in this queue immediately after applying any other kind of diff. *)
+    ; root_transitions: Diff.Root_transition.Lite.t Queue.t }
+
+  type nonrec create_args = create_args
 
   type output = unit
 
   (* worker assumes database has already been checked and initialized *)
-  let create = Fn.id
+  let create ({db; logger} : create_args) : t =
+    {db; logger; root_transitions= Queue.create ()}
+
+  let eagerly_perform_root_transitions t =
+    let start = Time.now () in
+    let finish count =
+      [%log' trace t.logger] "Eagerly performed $n root transitions in $time"
+        ~metadata:
+          [ ("n", `Int count)
+          ; ("time", `String Time.(Span.to_string_hum (diff (now ()) start)))
+          ]
+    in
+    let rec go count =
+      match Queue.peek t.root_transitions with
+      | None ->
+          finish count
+      | Some {new_root; garbage} -> (
+          let garbage =
+            match garbage with
+            | Lite garbage ->
+                garbage
+            | Full _ ->
+                assert false
+          in
+          match Database.move_root t.db ~new_root ~garbage with
+          | Ok _old_root ->
+              ignore (Queue.dequeue_exn t.root_transitions) ;
+              go (count + 1)
+          | Error _ ->
+              finish count )
+    in
+    go 0
```

Comment:
Mild nitpick: can you rewrite this as `let count = go 0 in (* contents of finish here *)`? I think it's a little easier reason about the control flow then (don't have to check that all terminating branches of `go` call `finish`).